opm/autodiff/BlackoilWellModel_impl.hpp

```diff
@@ -643,53 +671,63 @@ namespace Opm {
-        initPrimaryVariablesEvaluation();
+            updateWellControls();
+            // Set the well primary variables based on the value of well solutions
+            initPrimaryVariablesEvaluation();
 
-        if (param_.solve_welleq_initially_ && iterationIdx == 0) {
-            // solve the well equations as a pre-processing step
-            last_report_ = solveWellEq(dt, local_deferredLogger);
+            if (param_.solve_welleq_initially_ && iterationIdx == 0) {
+                // solve the well equations as a pre-processing step
+                    last_report_ = solveWellEq(dt, local_deferredLogger);
 
-            if (initial_step_) {
-                // update the explicit quantities to get the initial fluid distribution in the well correct.
-                calculateExplicitQuantities();
-                prepareTimeStep(local_deferredLogger);
-                last_report_ = solveWellEq(dt, local_deferredLogger);
-                initial_step_ = false;
+                if (initial_step_) {
+                    // update the explicit quantities to get the initial fluid distribution in the well correct.
+                    calculateExplicitQuantities();
+                    prepareTimeStep(local_deferredLogger);
+                    last_report_ = solveWellEq(dt, local_deferredLogger);
+                    initial_step_ = false;
+                }
+                // TODO: should we update the explicit related here again, or even prepareTimeStep().
+                // basically, this is a more updated state from the solveWellEq based on fixed
+                // reservoir state, will tihs be a better place to inialize the explict information?
             }
-            // TODO: should we update the explicit related here again, or even prepareTimeStep().
-            // basically, this is a more updated state from the solveWellEq based on fixed
-            // reservoir state, will tihs be a better place to inialize the explict information?
+
+            assembleWellEq(dt, local_deferredLogger);
+        } catch (...) {
+            exception_thrown = 1;
         }
-        assembleWellEq(dt, local_deferredLogger);
+        check_for_exceptions_and_log_and_throw(local_deferredLogger, exception_thrown, "solveWellEq in BlackoilWellModel::assemble() failed.", terminal_output_);
 
         Opm::DeferredLogger global_deferredLogger = gatherDeferredLogger(local_deferredLogger);
         if (terminal_output_) {
             global_deferredLogger.logMessages();
         }
 
-
         last_report_.converged = true;
     }
 
-
-
-
-
     template<typename TypeTag>
     void
     BlackoilWellModel<TypeTag>::
     assembleWellEq(const double dt, Opm::DeferredLogger& deferred_logger)
     {
+        int exception_thrown = 0;
         for (auto& well : well_container_) {
-            well->assembleWellEq(ebosSimulator_, dt, well_state_, deferred_logger);
+            try {
```

Comment:
Can the try be moved outside the for loop?

Reply:
done